### DISCUSSION TITLE
Typescript ES2020 target is not supported in older typescript compilers

### DISCUSSION THREAD

**Discussion by igorz-** [panva/oauth4webapi#110] [Q&A]
I choose oauth4webapi for one of our older projects at work primarily because it's a low-level framework agnostic library.
So I expected it shouldn't be a problem to use in a project that is built with outdated technologies.

Unfortunately, I encountered an error:
```ERROR in ./node_modules/oauth4webapi/build/index.js
Module parse failed: Unexpected token (2:61)
You may need an appropriate loader to handle this file type.
| let USER_AGENT;
| if (typeof navigator === 'undefined' || !navigator.userAgent?.startsWith?.('Mozilla/5.0 ')) {
```

So my question: Is there a reason why ES2020 chosen as a compilation target, whereas it reduces compatibility without providing significant benefits?
Would it be possible to change the compilation target to ES5 (or something more compatible).
Or maybe I'm missing something and the problem is not with typescript?

P.S. I use typescript 2.7

**Comment by panva** [comment]
I'm using ES2020 target mostly because of optional chaining, the very thing your build step complains about. Going older than ES2020 by a few steps means no Object.fromEntries, Array.prototype.includes, going even older than that then becomes complete insanity.

The compilation target is even in a comfortable place wrt. browser install/support base in relation to when their versions supporting the features needed were released.

Going any further with ES compatibility you can certainly achieve, but as part of your build steps.

**Reply by igorz-** [reply]
> I'm using ES2020 target mostly because of optional chaining

Optional chaining is actually supported in Typescript regardless of compilation target: https://www.typescriptlang.org/play?target=1&module=5#code/MYewdgzgLgBAhgLhgbwA4CcSoIIH4lqaoBC+MYArgLYBGApugL6MwC8KjA3AFDeiSwabeADoMWPGKLEeQA

> Going any further with ES compatibility you can certainly achieve, but as part of your build steps.

Unfortunately, this requires upgrading our build tools.

If it's not possible to set the target to maybe ES6 for this library, I'll find another way.

P.S.: My two cents is making a library more compatible can increase its popularity. Considering that this library is quite low-level I assume that the target audience may want to use it in a variety of different environments.
